File: base/scripts/containers/debug_panel.lua
class "DebugPanel" (Dialog)

function DebugPanel:init()
    self.fps_lbl = Label("fpsLbl")
    self.fps_lbl:setFont("system_15_fnt")
    self.fps_lbl:setTextAlignment("CENTER|MIDDLE")
    self.fps_lbl:setRect(0, 0, 100, 30)
    self.fps_lbl:setColour("green")
    self.fps_lbl:setAlignment("LEFT|TOP", 0, 0)
    self.fps_lbl:setText("0")
    self:attach(self.fps_lbl)

    self.mouse_pos_lbl = Label("mousePosLbl")
    self.mouse_pos_lbl:setFont("system_15_fnt")
    self.mouse_pos_lbl:setTextAlignment("CENTER|MIDDLE")
    self.mouse_pos_lbl:setRect(0, 0, 120, 30)
    self.mouse_pos_lbl:setColour("green")
    self.mouse_pos_lbl:setAlignment("RIGHT|TOP", 0, 0)
    self.mouse_pos_lbl:setText("0")
    self:attach(self.mouse_pos_lbl)

    self.timer = Timer("updateTimer")
    self.timer:restart(100)
    self.timer:addCallback("TimerElapsed", self.onTimerElapsed, self)
    self:attach(self.timer)
end

local mouse_pos_format = "%s|%s"

function DebugPanel:onTimerElapsed()
    if (self.fps_lbl) then
        self.fps_lbl:setText(Engine.getFPS())
    end
    if (self.mouse_pos_lbl) then
        local x, y = Engine.getMousePos()
        self.mouse_pos_lbl:setText(string.format(mouse_pos_format, x, y))
    end
    if (self.timer) then
        self.timer:restart(100)
    end
end

File: src/debug_panel.cpp
#include "debug_panel.h"

namespace redrevolt
{
DebugPanel::DebugPanel(const std::string & id)
    : stren::Dialog(id)
    , m_fpsLabel(nullptr)
    , m_mousePosLabel(nullptr)
    , m_timer(nullptr)
{
    stren::Label * fpsLabel = new stren::Label("fpsLbl");
    fpsLabel->setFont("system_15_fnt");
    fpsLabel->setTextAlignment("CENTER|MIDDLE");
    fpsLabel->setRect(0, 0, 100, 30);
    fpsLabel->setColour("green");
    fpsLabel->setAlignment("LEFT|TOP", 0, 0);
    fpsLabel->setText("0");
    attach(fpsLabel);
    m_fpsLabel = fpsLabel;

    stren::Label * mousePosLabel = new stren::Label("mousePosLbl");
    mousePosLabel->setFont("system_15_fnt");
    mousePosLabel->setTextAlignment("CENTER|MIDDLE");
    mousePosLabel->setRect(0, 0, 120, 30);
    mousePosLabel->setColour("green");
    mousePosLabel->setAlignment("RIGHT|TOP", 0, 0);
    mousePosLabel->setText("0");
    attach(mousePosLabel);
    m_mousePosLabel = mousePosLabel;

    stren::Timer * timer = new stren::Timer("updateTimer");
    timer->restart(100);
    timer->addCallback("TimerElapsed", this, &DebugPanel::onTimerElapsed);
    attach(timer);
    m_timer = timer;
}

DebugPanel::~DebugPanel()
{
}

void DebugPanel::onTimerElapsed(Widget * sender)
{
    if (m_fpsLabel)
    {
        char str[16];
        sprintf(str, "%d", stren::Engine::getFPS());
        m_fpsLabel->setText(str);
    }
    if (m_mousePosLabel)
    {
        const stren::Point & pos = stren::Engine::getMousePos();
        char str[16];
        sprintf(str, "%d|%d", pos.getX(), pos.getY());
        m_mousePosLabel->setText(str);
    }
    if (m_timer)
    {
        m_timer->restart(100);
    }
}
} // redrevolt
